common4j/src/main/com/microsoft/identity/common/java/util/ported/Pair.java

```diff
@@ -23,10 +23,54 @@
 
 package com.microsoft.identity.common.java.util.ported;
 
+import java.util.List;
+import java.util.Objects;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 import lombok.AllArgsConstructor;
+import lombok.NonNull;
 
 @AllArgsConstructor
 public class Pair<T, U> {
     public final T first;
     public final U second;
+
+    @Override
+    @SuppressFBWarnings(value = "NP_METHOD_PARAMETER_TIGHTENS_ANNOTATION",
+            justification = "Equals function should be able to take in a null object.")
+    public boolean equals(@Nullable final Object obj) {
+        if (obj == null){
+            return false;
+        }
+
+        if (!(obj instanceof Pair)) {
+            return false;
+        }
+
+        final Pair<?, ?> castedObj = (Pair<?, ?>) obj;
+        return Objects.equals(castedObj.first, first) && Objects.equals(castedObj.second, second);
+    }
+
+    /**
+     * Compute a hash code using the hash codes of the underlying objects
+     *
+     * @return a hashcode of the Pair
+     */
+    @Override
+    public int hashCode() {
+        int result = (first != null ? first.hashCode() : 0);
+        result = 31 * result + (second != null ? second.hashCode() : 0);
```

Comment:
Using `31 * result`  here will cause these pairs to have different hashcodes,
```java
Pair<String, String> first = new Pair<>("foo", "bar");
Pair<String, String> second = new Pair<>("bar", "foo");
```

But structurally, they have the same pair of values. Is this intentional, and if so, what's the reason for considering them as different pairs?

Reply:
It's intentional.

We're using pair to store query parameter. Let's say brkr=1

brkr=1 should not be equal to 1=brkr.

Reply:
(Maybe I should've renamed this to KeyValuePair)

Reply:
Yeah, it's more of a `KeyValuePair` than a `Pair`.

Reply:
Use lombok?

Reply:
@shahzaibj just addressed with EqualsAndHashCode as @AdamBJohnsonx  suggested above.